src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs

```diff
@@ -237,7 +248,12 @@ public bool TryRemoveDocumentFromWorkspace(Workspace workspace, string filePath)
         {
             if (_fileToDocumentMap.TryGetValue(filePath, out var documentId))
             {
-                workspace.OnDocumentClosed(documentId, new FileTextLoader(filePath, MetadataAsSourceGeneratedFileInfo.Encoding));
+                var document = workspace.CurrentSolution.GetDocument(documentId);
+                // This method is ultimately called from an RDT event on the UI thread, so we get the text
+                // synchronously. We need the text to preserve the encoding.
+                var encoding = document?.GetTextSynchronously(CancellationToken.None).Encoding;
+
+                workspace.OnDocumentClosed(documentId, new FileTextLoader(filePath, encoding ?? Encoding.UTF8));
```

Comment:
I would have assumed the "encoding" here would be the default/fallback encoding that was originally used when the project was loaded earlier; do we still have that around somewhere?

Reply:
That's what this is doing, pulling the encoding out of the document from when we created it above. Or do you mean we should cache the encoding with the project to save looking up the text?

Reply:
The latter.

Reply:
👍  Updated